Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,414 @@
+            private Vector3 m_boundingShapeScale;
+            private Quaternion m_boundingShapeRotation;
+            
+            public Collider2D m_boundingShape2D;
+            public Vector2 m_boundingShapeOffset;
+            public List<List<Vector2>> m_originalPath;
+            public List<ConfinerOven.ConfinerState> m_confinerStates;
+
+            public void Invalidate()
+            {
+                m_aspectRatio = 0;
+                m_maxOrthoSize = 0;
+                m_shrinkToPoints = false;
+                
+                m_boundingShapePosition = Vector3.negativeInfinity;
+                m_boundingShapeScale = Vector3.negativeInfinity;
+                m_boundingShapeRotation = new Quaternion(0,0,0,0);
+                
+                m_boundingShape2D = null;
+                m_originalPath = null;
+
+                m_confinerStates = null;
+            }
+
+            public bool IsValid(in Collider2D boundingShape2D, 
+                in float aspectRatio, in float maxOrthoSize, in bool shrinkToPoint)
+            {
+                return m_boundingShape2D != null && m_boundingShape2D == boundingShape2D && // same boundingShape?
+                       !BoundingShapeTransformChanged(boundingShape2D.transform) && // input shape changed?
+                       m_boundingShapeOffset == boundingShape2D.offset && // same offset on boundingShape?
+                       m_originalPath != null && // first time?
+                       m_confinerStates != null && // cache not empty? 
+                       Mathf.Abs(m_aspectRatio - aspectRatio) < UnityVectorExtensions.Epsilon && // aspect changed?
+                       Mathf.Abs(m_maxOrthoSize - maxOrthoSize) < UnityVectorExtensions.Epsilon && // max ortho changed?
+                       m_shrinkToPoints == shrinkToPoint; // shrink to point option changed
+            }
+
+            public void SetTransformCache(in Transform boundingShapeTransform)
+            {
+                m_boundingShapePosition = boundingShapeTransform.position;
+                m_boundingShapeScale = boundingShapeTransform.localScale;
+                m_boundingShapeRotation = boundingShapeTransform.rotation;
+            }
+            
+            private bool BoundingShapeTransformChanged(in Transform boundingShapeTransform)
+            {
+                return m_boundingShape2D != null && 
+                       (m_boundingShapePosition != boundingShapeTransform.position ||
+                        m_boundingShapeScale != boundingShapeTransform.localScale ||
+                        m_boundingShapeRotation != boundingShapeTransform.rotation);
+            }
+        }
+        private ShapeCache m_shapeCache;
+
+        /// <summary>
+        /// VcamShapeCache (lives inside VcamExtraState): contains all the cache items
+        /// that are dependent on something in the vcam.
+        /// </summary>
+        private struct VcamShapeCache
```

Comment:
I meant actually have this definition live inside VcamExtraState - it will make naming simpler